README.md
# Arduino Weather Station

This is some code to enable connecting a Davis anemometer directly to an Arduino, without the Davis base station (aka the "ISS") or the console. The Arduino then sends the wind data to whatever computer it's attached to as serial data. I also wrote a driver to use it with my favorite weather station proram, weeWX. So connect it to your Raspberry Pi or whatever and you get a pro quality internet connected wind gauge for about $130.

The anemometer I used is sold as a replacement part for the Davis Vantage Pro2. They're on eBay in the $110 range. The one I purchased had the auction name "Davis Instruments 6410 Davis Anemometer For Vantage Pro Pro2 6152". 

Here's a test page I put up that's receiving data from it every few seconds:

http://rproductions.org/wind


## Wiring and Setup

Here's a wiring diagram to connect the Davis anemometer to an Arduino thanks to the good folks at [cactus.io](http://cactus.io/hookups/weather/anemometer/davis/hookup-arduino-to-davis-anemometer):

![alt tag](https://github.com/wrybread/ArduinoWeatherStation/raw/master/arduino-to-davis-anemometer-hookup-circuit.jpg)

Or use their instructions directly, since there's a bit more detail on their site:

http://cactus.io/hookups/weather/anemometer/davis/hookup-arduino-to-davis-anemometer

A nice way to wire up the anemometer is to use a phone cable extension chord or coupler so you don't need to splice the line coming from the anemometer. Works great but don't forget to flip the wire layout in the diagram above. I unfortunately had to learn that one the hard way. But the good news is that there appears to be some reverse polarity protection in the anemometer...

For the Arduino sketch, see ArduinoWeather.ino above, which I adapted from the code [at cactus.io](http://cactus.io/hookups/weather/anemometer/davis/hookup-arduino-to-davis-anemometer):

Then simply connect the Arduino to the USB port of your Pi or whatever. Note the script arduino_test.py, which should start printing the data being sent by your Arduino. If not, you might have to set your port at the top of the file.

To use this with the weeWX weather station program, see the driver aws.py. I need to package it with the weeWX extension manager to make installation a bit easier.


## Installation 

- load ArduinoWeather.ino onto your Arduino

- install the AWS weeWX driver included here (aws.py) in your user directory. It *should* be possible to use the wee_extension tool to install it (see below for instructions), but I haven't confirmed that yet. Anyway, copying the driver (aws.py) to weewx/bin/user/aws.py should do it.

- add this section to your weeWX config file:

[AWS]

    # This section is for an Arduion Weather Station.

    # Serial port such as /dev/ttyACM0, /dev/ttyS0, /dev/ttyUSB0, or /dev/cuaU0
    port = /dev/ttyACM0

    # The driver to use:
    driver = user.aws

- in the [Station] section of your weeWX config file, set "station_type" to "AWS".



## To do:

- set up the Arduino Weather Station next to a Davis Vantage Pro2 and compare the readings

- test installation using the wee_extension utility. These instructions should work:

1) Download the driver:

wget -O weewx-aws.zip https://github.com/wrybread/weewx-ArduinoWeatherStation.git

2) Install the driver:

sudo wee_extension --install weewx-aws.zip

3) Configure the driver:

sudo wee_config --reconfigure


changelog
0.49 20mar2017
* fix rain counter bug

0.48 18mar2017
* deal with bogus characters in readings

0.47 22jan2017
* user-specified map augments the default instead of replacing it

0.46 07nov2016
* use the database_field = hardware_name pattern for sensor map

0.45 29sep2016
* eliminate unecessary unit conversions
* simplify some code

0.44 29sep2016
* Fixed: output machine readable values in US format
* Added: automatic selection of digital and analog temp sensors
* Fixed: logic for analog temp sensors

0.43 17sep2016
* Fixed: logic for outside temperature readings

0.42 31aug2016
* Fixed: logic for raw signals of temp/hum stations
* Changed: formula for wind_dir_raw
* Changed: pct_good not sent to weewx at first (not full) archive period 
* Added: rf_sensitivity too low warnings

0.41 08jul2016
* Added: custom receiver inHumidity support and other small fixes
* Added: wind error correction code and associated metric system change
* Fixed: Solar "sensor missing" fix by Luc

0.40 09jun2016
* Added: routine new_archive_record 
* Changed: handling of rf_stats moved to new_archive_record
* Changed: SM-MAP and LW_map settings
* Changed: all debug levels from 1 to 2
* Changed: removed rainRate from DEFAULT_SENSOR_MAP
* Changed: packets with only a battery status are now skipped
* Changed: made handling of rain_count more robust
* Changed: pctgood is now rounded to the nearest integer value
* Added: Raw rainrate value checked for no sensor
* Fixed: battery status for leaf-soil station was missing
* Changed: Removed debug code in leaf-soil section 

0.39 31may2016
* Change: replaced calculate_soil_moisture by lookup_potential
* Change: replaced calculate_leaf_wetness by lookup_potential
* Change: no debug message if rain_count=128 (128=no sensor)
* Change: set wind_dir to None when wind_speed == 0 (or < 0.001)
* Added: set leaf_temp = soil_temp in machine format for channels 1 and 2
* Change: soil moisture potential calulated with norm factor 0.009
* Change: leaf wetness potential calulated with norm factor 0.0

0.38 31may2016
* make the initialization process more robust

0.37 25may2016
* Changed: calculate_soil_moisture
* Changed: calculate_leaf_wetness
* Changed: tag for solar_power.
* Added: tag for supercap_volt.
* Fixed: formula for supercap voltage raw value (Vue only).
* Fixed: formula for uv raw value
* Fixed: typo with light rain
* Fixed: accuracy rain rate formula's
* Fixed: formula for solar cell output raw value
* Changed: explanation of unknown ATK message (0xC)
* Changed: the value of not connected soil temperature sensors
* Changed: the value of not connected soil moisture sensors
* Changed: the value of not connected leaf wetness sensors

0.36 12may2016
* Changed: logic for soil moisture
* Changed: logic for leaf wetness
* Fixed: formula for outtemp
* Added: message examples
* Fixed: handle TypeError when pctgood is None
* Fixed: data of not-present leaf wetness sensors can have values 001 or 002
* Changed: some logdbg messages into log_parse messages
* Changed: some texts

0.35 07may2016
* Changed the way pct_good and pctgood are handled and calculated
* Instead of METEOSTICK_CHANNEL we now use RAW_CHANNEL and MACHINE_CHANNEL
* Changed the time when update_rf_summaries is run
* Fixed self.output_format = fmt.lower()
* Fixed missing 'else' in parse
* Added battery low stati for output raw
* Changed dbg_parse levels
* Unused leaf wetness sensor has raw value 0x002 where most others have 0x3FF
* Changed coding in leaf-soil section

0.34 06may2016
* default to 'raw' format not 'machine' format
* rf stats are recorded only if debug_rf is enabled
* push parameters down to meteostick class
* keep wx calculations out of the driver and hardware code
* incorporated luc's transitional soil calculations
* rain_rate is now in loop packets, but intentionally not in the sensor map

0.33 29apr2016
* default to raw format

0.32 27apr2016
* fixed bug in decode_soil
* checks to see if raw data not present
* change soil moisture formula
* added raw values of leaf_wetness to observations
* added leaf_temp formula
* minor log formatting changes

0.31 26apr2016
* added parsing of raw output
* added soil calculations

0.15 24apr2016
* ignore (and log) invalid rain count values
* track pct_good stats with rf stats
* fix log level of parsing messages
* added configurator for setting/getting meteostick parameters
* consolidate rf stat calculations
* eliminate redundant cmd sequences

0.14 23apr2016
* added RF signal tracking

0.13 20apr2016
* default rain bucket is 1 to match default frequency of EU
* added flag for debugging RF sensitivity

0.12 20apr2016
* added option to set RF sensitivity
* added support for australian frequencies

0.11 13apr2016
* use int for rain_count

0.10 12apr2016
* rain counter wraparound is 128 not 256
* prevent failure when empty sentence from meteostick
* catch parsing exceptions so that failures will not cause weewx to quit
* adjust log messages during startup

0.9 11apr2016
* always provide details about meteostick state

0.8 09apr2016
* minor code refactoring
* make default config stanza consistent
* individual indicators for each battery status
* remove wview-isms from sensor observation names
* make rain variables consistent

0.6 23mar2016
* added parameters for bucket type, transceiver frequency, and channels
* added support for soil moisture, extra temp, extra humid

0.4 19feb2016
* added mapping for rf_signal_1 to rxCheckPercent

0.3 18feb2016
* no log of field rf_signal_*
* no log for blank lines
* lines starting with # will be logged as info message
* changed format of B

0.1 10feb2016
* initial release


arduino_test.py
#!/usr/bin/env python


'''

Simple script that reads serial data from the Arduino attached to whatever port.


'''


# adjust this to taste
port = "/dev/ttyACM0"





import time, os, serial 


print "Connecting to %s..." % port
ser = serial.Serial(port, 9600, timeout=.5)

print "Success!"



def read_buffer():

    # read the whole buffer
    last_data=''
    while True:
        data=ser.readline()
        if data!='':
            last_data=data
        else:
            return last_data.replace("\n", "")





while True:

    data = read_buffer()

    if data:
        print data

    



aws.py
#!/usr/bin/env python
#
# This is a weeWX driver to enable weeWX to read data from an Arduino.
#
# See here for more details:
#
# https://github.com/wrybread/ArduinoWeatherStation
#
# by wrybread@gmail.com
#
#
#



"""Driver for Arduion Weather Station.

See here for more info:

https://github.com/wrybread/ArduinoWeatherStation

"""

from __future__ import with_statement
import serial
import syslog
import time

import weewx.drivers

DRIVER_NAME = 'AWS'
DRIVER_VERSION = '0.1'


def loader(config_dict, _):
    return AWSDriver(**config_dict[DRIVER_NAME])

def confeditor_loader():
    return AWSConfEditor()


INHG_PER_MBAR = 0.0295333727
METER_PER_FOOT = 0.3048
MILE_PER_KM = 0.621371

DEFAULT_PORT = '/dev/ttyACM0'
DEBUG_READ = 0






def logmsg(level, msg):
    syslog.syslog(level, 'aws: %s' % msg)

def logdbg(msg):
    logmsg(syslog.LOG_DEBUG, msg)

def loginf(msg):
    logmsg(syslog.LOG_INFO, msg)

def logerr(msg):
    logmsg(syslog.LOG_ERR, msg)







class AWSDriver(weewx.drivers.AbstractDevice):
    """weewx driver that communicates with an Arduino weather station
    
    port - serial port
    [Required. Default is /dev/ttyACM0]

    polling_interval - how often to query the serial interface, seconds
    [Optional. Default is 1]

    max_tries - how often to retry serial communication before giving up
    [Optional. Default is 5]

    retry_wait - how long to wait, in seconds, before retrying after a failure
    [Optional. Default is 10]
    """
    def __init__(self, **stn_dict):
        self.port = stn_dict.get('port', DEFAULT_PORT)
        self.polling_interval = float(stn_dict.get('polling_interval', 1))
        self.max_tries = int(stn_dict.get('max_tries', 5))
        self.retry_wait = int(stn_dict.get('retry_wait', 10))
        self.last_rain = None
        loginf('driver version is %s' % DRIVER_VERSION)
        loginf('using serial port %s' % self.port)
        loginf('polling interval is %s' % self.polling_interval)
        global DEBUG_READ
        DEBUG_READ = int(stn_dict.get('debug_read', DEBUG_READ))

        self.last_read_time = time.time()
        self.read_counter = 0


        logdbg("Opening the Arduino on port %s" % self.port)

        self.baudrate = 9600
        self.timeout = 10 # changed from 60
        self.serial_port = None
        
        self.serial_port = serial.Serial(self.port, self.baudrate,
                                         timeout=self.timeout)


        print "done initting" #%%



    def read_buffer(self):

        # read the buffer
        new_data = ""

        try:
            while True:
                new_data = self.serial_port.readline()
                new_data = new_data.replace("\n", "")
                return new_data

        except Exception, e:
            pass



    def parse_readings(self, b):
        """
	  The Arduino script emits data in the format:
	  
          [0]wind speed
          [1]wind direction
          [2]wind direction compass
          [3]temperature
          [4]barometer
        """
        
        parts = b.split(",")
        
        #print "parsing: ", b, "parts =", parts
        
        data = dict()
        
        try:
            data['windSpeed'] = float(parts[0])  # mph
            data['windDir'] = float(parts[1])

            # these might have to be renamed
            #data['temperature'] = float(parts[2])
            #data['barometer'] = float(parts[3])
            
            #data['windDirCompass'] = parts[2]  # wind speed compass. Unused, or probably wrong variable name.

        except Exception, e:
            logerr("Error parsing data: %s" % e)
        
        #%%
        print "aws: ", data

        if DEBUG_READ:
            logdbg(data)

            
        return data


        


    @property
    def hardware_name(self):
        return "AWS"


    def genLoopPackets(self):

        ntries = 0
        
        while ntries < self.max_tries:
            
            ntries += 1
            
            try:
                
                packet = {'dateTime': int(time.time() + 0.5),
                          'usUnits': weewx.US}
                
                serial_data = self.read_buffer() # read the buffer from the Arduino
                    
                data = self.parse_readings(serial_data) # parse the data
                
                packet.update(data) 
                
                self._augment_packet(packet)
                
                ntries = 0


                '''
                # print the time between reads and the count for debugging for now
                self.read_counter += 1
                time_since_last_read = time.time() - self.last_read_time
                print "%s seconds since last read" % time_since_last_read, self.read_counter
                self.last_read_time = time.time()
                '''
                
                yield packet

                if self.polling_interval:
                    time.sleep(self.polling_interval)
                    
            except (serial.serialutil.SerialException, weewx.WeeWxIOError), e:
                logerr("Failed attempt %d of %d to get LOOP data: %s" %
                       (ntries, self.max_tries, e))
                time.sleep(self.retry_wait)
                
        else:
            msg = "Max retries (%d) exceeded for LOOP data" % self.max_tries
            logerr(msg)
            raise weewx.RetriesExceeded(msg)



    def _augment_packet(self, packet):

        # no wind direction when wind speed is zero
        if 'windSpeed' in packet and not packet['windSpeed']:
            packet['windDir'] = None











class AWSConfEditor(weewx.drivers.AbstractConfEditor):
    
    @property
    def default_stanza(self):
        return """
[AWS]
    # This section is for an Arduion Weather Station.

    # Serial port such as /dev/ttyACM0, /dev/ttyS0, /dev/ttyUSB0, or /dev/cuaU0
    port = /dev/ttyACM0

    # The driver to use:
    driver = user.aws
"""

    def prompt_for_settings(self):
        print "Specify the serial port on which the station is connected, for"
        print "example /dev/ttyUSB0 or /dev/ttyACM0."
        port = self._prompt('port', '/dev/ttyACM0')
        return {'port': port}
















    
# define a main entry point for basic testing of the station without weewx
# engine and service overhead.  invoke this as follows from the weewx root dir:
#
# To run on a standard apt-get install (unconfirmed, might ened to adjust paths):
# PYTHONPATH=bin python /home/weewx/bin/user/aws.py

# For setup.py instals:
# PYTHONPATH=/home/weewx/bin python /home/weewx/bin/user/aws.py
#

if __name__ == '__main__':
    
    import optparse

    usage = """%prog [options] [--help]"""

    syslog.openlog('aws', syslog.LOG_PID | syslog.LOG_CONS)
    syslog.setlogmask(syslog.LOG_UPTO(syslog.LOG_DEBUG))
    parser = optparse.OptionParser(usage=usage)
    parser.add_option('--version', dest='version', action='store_true',
                      help='display driver version')
    parser.add_option('--port', dest='port', metavar='PORT',
                      help='serial port to which the station is connected',
                      default=DEFAULT_PORT)
    (options, args) = parser.parse_args()

    if options.version:
        print "AWS driver version %s" % DRIVER_VERSION
        exit(0)



    station = AWSDriver()
    for packet in station.genLoopPackets():
        print time.time(),  packet

    
